performance/src/main/java/org/hibernate/validator/performance/simple/MultiLevelContainerValidation.java

```diff
@@ -0,0 +1,273 @@
+			Map<Cinema, List<EmailAddress>> map = new LinkedHashMap<>();
+			for ( int i = 0; i < maxEntries; i++ ) {
+				map.put( Cinema.get(), EmailAddress.getList( maxListEntries ) );
+			}
+			return map;
+		}
+
+		private static String randomString() {
+			char[] chars = new char[RANDOM.nextInt( 10 ) + 1];
+			for ( int i = 0; i < chars.length; i++ ) {
+				chars[i] = (char) RANDOM.nextInt();
+			}
+			return String.valueOf( chars );
+		}
+
+		/**
+		 * Model classes.
+		 */
+
+		static class MapContainer {
+
+			private Map<@NotNull Optional<@Valid Cinema>, List<@NotNull @Valid EmailAddress>> map;
+
+			public MapContainer(Map<@NotNull Optional<@Valid Cinema>, List<@NotNull @Valid EmailAddress>> map) {
+				this.map = map;
+			}
+
+			public Map<Optional<Cinema>, List<EmailAddress>> getMap() {
+				return map;
+			}
+		}
+
+		static class UnConstraintMapContainer {
+
+			private Map<Optional<Cinema>, List<EmailAddress>> map;
+
+			public UnConstraintMapContainer(Map<Optional<Cinema>, List<EmailAddress>> map) {
+				this.map = map;
+			}
+
+			public Map<Optional<Cinema>, List<EmailAddress>> getMap() {
+				return map;
+			}
+		}
+
+		static class OtherMapContainer {
+
+			private Map<@NotNull @Valid Cinema, List<EmailAddress>> map;
+
+			public OtherMapContainer(Map<@NotNull @Valid Cinema, List<EmailAddress>> map) {
+				this.map = map;
+			}
+
+			public Map<Cinema, List<EmailAddress>> getMap() {
+				return map;
+			}
+		}
+
+		static class Cinema {
+
+			String name;
+
+			Reference<@Valid Visitor> visitor;
+
+			Cinema(String name, Reference<Visitor> visitor) {
+				this.name = name;
+				this.visitor = visitor;
+			}
+
+			static Cinema get() {
+				return new Cinema( randomString(), new SomeReference<>( Visitor.get() ) );
+			}
+		}
+
+		interface Reference<T> {
+
+			T getValue();
+		}
+
+		static class SomeReference<T> implements Reference<T> {
+
+			private final T value;
+
+			public SomeReference(T value) {
+				this.value = value;
+			}
+
+			@Override
+			public T getValue() {
+				return value;
+			}
+		}
+
+		static class Visitor {
+
+			@NotNull
+			String name;
+
+			Visitor(String name) {
+				this.name = name;
+			}
+
+			static Visitor get() {
+				return new Visitor( randomString() );
+			}
+		}
+
+		static class EmailAddress {
+
+			// not using @Email as it will require HV as a dependency and these tests can be run against bVal as well
+			@NotNull
+			@Pattern(regexp = "(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])")
```

Comment:
I would say, just keep the `@NotNull` constraint. The point is to exercise the validator code, not to have an expensive constraint checked. I would have used `@NotBlank` for a change but it's not part of the spec either :).